src/Components/Web.JS/src/Boot.Web.ts

```diff
@@ -31,33 +30,59 @@ function boot(options?: Partial<WebStartOptions>) : Promise<void> {
   started = true;
 
   Blazor._internal.loadWebAssemblyQuicklyTimeout = 3000;
+
   // Defined here to avoid inadvertently imported enhanced navigation
   // related APIs in WebAssembly or Blazor Server contexts.
   Blazor._internal.hotReloadApplied = () => {
-    if (hasProgrammaticEnhancedNavigationHandler())
-    {
+    if (hasProgrammaticEnhancedNavigationHandler()) {
       performProgrammaticEnhancedNavigation(location.href, true);
     }
-  }
+  };
 
   setCircuitOptions(options?.circuit);
   setWebAssemblyOptions(options?.webAssembly);
 
-  rootComponentManager = new WebRootComponentManager(options?.ssr?.circuitInactivityTimeoutMs ?? 2000);
+  const rootComponentManager = new WebRootComponentManager(options?.ssr?.circuitInactivityTimeoutMs ?? 2000);
+  const enqueueDispatchEnhancedLoad = createEnhancedLoadDispatcher();
+
+  const navigationEnhancementCallbacks: NavigationEnhancementCallbacks = {
+    documentUpdated: () => {
+      rootComponentManager.onDocumentUpdated();
+      enqueueDispatchEnhancedLoad();
+    },
+  };
 
   attachComponentDescriptorHandler(rootComponentManager);
-  attachStreamingRenderingListener(options?.ssr, rootComponentManager);
+  attachStreamingRenderingListener(options?.ssr, navigationEnhancementCallbacks);
 
   if (!options?.ssr?.disableDomPreservation) {
-    attachProgressivelyEnhancedNavigationListener(rootComponentManager);
+    attachProgressivelyEnhancedNavigationListener(navigationEnhancementCallbacks);
   }
 
   registerAllComponentDescriptors(document);
-  rootComponentManager.documentUpdated();
+  rootComponentManager.onDocumentUpdated();
 
   return Promise.resolve();
 }
 
+// This function ensures that 'enhancedload' only gets invoked once
+// for any synchronous sequence of document updates via SSR.
+function createEnhancedLoadDispatcher() {
+  let isDispatchPending = false;
+
+  return function() {
+    if (isDispatchPending) {
+      return;
+    }
+
+    isDispatchPending = true;
+    setTimeout(() => {
+      isDispatchPending = false;
+      Blazor._internal.dispatchEvent('enhancedload', {});
+    }, 0);
```

Comment:
`setTimeout(..., 0)` always sets off alarms since it means the caller loses the ability to do anything synchronous or even run any logic after this callback, without reverting to even more problematic tricks like `setTimeout(..., 1)` and so on.

Not saying it's necessarily wrong since maybe this really is the only/cleanest way to achieve the goal. But could you clarify why SSR will do things multiple times synchronously, and whether we could instead make SSR responsible for only triggering this once at the end of its process?

If this comes down to "a single chunk of HTTP traffic may include multiple `<blazor-ssr>` blocks" then sure that's true, but they may arrive arbitrarily quickly anyway, so it might not be particularly useful to distinguish "those that arrive synchronously together" from those who don't. What difference would it make if we fired the event for each one, whether or not there are others in the same synchronous block? The recipient of the event can still debounce if it needs to for its own reasons.

Reply:
> Not saying it's necessarily wrong since maybe this really is the only/cleanest way to achieve the goal. But could you clarify why SSR will do things multiple times synchronously, and whether we could instead make SSR responsible for only triggering this once at the end of its process?

We currently invoke this callback:
* After the initial page load
* After the initial content is received during enhanced navigation
* After the document completes a streaming response update
* After the enhanced navigation completes
  * This might seem redundant, but it enables reacting to "quiescence". Blazor itself uses this so that new components get activated only when enhanced navigation completes.

Given that the public event (`'enhancedload'`) doesn't distinguish between these four cases, the `setTimeout()` mechanism was used to prevent, for example, the callback being invoked after receiving initial enhanced navigation content, then immediately again if there's no additional streaming content.

But I do agree that the recipient can debounce these events if it needs to, and not having the `setTimeout()` also allows us to extend this in the future if we expose which of the four cases resulted in the callback getting invoked (maybe someone wants to listen explicitly for the "navigation end" callback, and deduplicating this event results in that callback getting shadowed by the "initial load" callback).

Reply:
I've actually just updated this so the fourth case ("after the enhanced navigation completes") does not dispatch the `'enhancedload'` event, because it doesn't signal that changes to the DOM were made in that case. In the future, we could add a `navigationend` event to allow developers to react to that case.